### PR TITLE
Add support for reading CSV files with comments

### LINKED ISSUE #10262: Support skipping comments in `CsvReader`
It would be great if datafusion had out-of-the-box support for skipping comment lines. While non of this is "standardized" many CSV readers support skipping full comment lines. An often used comment indentifier is a `#` prefix (default in e.g. [pandas] or [R]).

I originally posted this as a comment https://github.com/apache/datafusion/issues/8824#issuecomment-2078835858. 

[pandas]: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html
[R]: https://stat.ethz.ch/R-manual/R-devel/library/utils/html/read.table.html

### PR DESCRIPTION
**Pull request by bbannier**
This PR adds support for parsing CSV files containing comment lines.

Closes #10262.

### CODE REVIEW

**Comment by bbannier** [comment]
This is currently a sketch for a possible implementation for #10262. The approach taken push interpretation of comment lines into `arrow-csv` apache/arrow-rs#5759 adding support for that; the task here is then to plug a `datafusion` comment config setting through to `arrow-csv`.

If this is a viable solution it would require a bump of at least the `arrow-csv` dependency in `datafusion` to a version containing support for comments. To at least explore that I prefixed the actual implementation patch here with two patches performing that bump. It appears that a bump to the `master` version of `arrow-csv` (or something else from the collection of crates in https://github.com/apache/arrow-rs) requires changes to `datafusion`; I attempted to perform that bump, but currently there are still some remaining issues,

```---- aggregates::tests::aggregate_source_not_yielding_with_spill stdout ----
Error: ResourcesExhausted("Failed to allocate additional 2208 bytes for GroupedHashAggregateStream[0] with 348 bytes already allocated - maximum available is 1600")

---- aggregates::tests::aggregate_source_with_yielding_with_spill stdout ----
Error: ResourcesExhausted("Failed to allocate additional 2208 bytes for GroupedHashAggregateStream[0] with 348 bytes already allocated - maximum available is 1600")

---- aggregates::tests::run_first_last_multi_partitions stdout ----
Error: ResourcesExhausted("Failed to allocate additional 3704 bytes for GroupedHashAggregateStream[0] with 437 bytes already allocated - maximum available is 3200")
```

@alamb, would you be open to shepherding this PR and https://github.com/apache/arrow-rs/pull/5759, or alternatively could identify someone who could?

**Comment by alamb** [comment]
> If this is a viable solution it would require a bump of at least the arrow-csv dependency in datafusion to a version containing support for comments. 

Yes. FWIW DataFusion typically upgrades to the latest arrow-rs (including arrow-csv) dependency so while extra time would be needed no extra work would be

**Comment by bbannier** [comment]
This is now rebased on `main` which recently bumped arrow.

**Review by alamb — COMMENTED** [review]
Thank you very much for this contribution @bbannier  -- this code looks great. The only thing I think this PR now needs is some test coverage so we don't break it in the future

Here is my suggestion for testing:
1. update [csv_files.slt](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/csv_files.slt), see [`this file`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/README.md) for info on running sql logic tests



Note I think you can programatically create a csv file with a command like

```sql
> copy (values ('column1,column2'), ('#second line is a comment'), ('2,3')) TO '/tmp/my.csv' OPTIONS ('format.delimiter' '|');
+-------+
| count |
+-------+
| 3     |
+-------+
1 row(s) fetched.
Elapsed 0.004 seconds.
```

That results in 
```bash
$ cat /tmp/my.csv
column1,column2
#second line is a comment
2,3
```

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 1]
```diff
@@ -202,3 +202,27 @@ select * from stored_table_with_necessary_quoting;
...
+CREATE EXTERNAL TABLE stored_table_with_comments (
+  c1 VARCHAR,
+  c2 VARCHAR
+) STORED AS CSV
+LOCATION 'test_files/scratch/csv_files/file_with_comments.csv'
+OPTIONS ('format.comment' '#',
+         'format.delimiter' ',');
+
+query TT
+SELECT * from stored_table_with_comments;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/csv_files.slt:225]: 👍 Love it

**Review by alamb — APPROVED** [review]
Thank you @bbannier 🚀

**Pull request merged**